app/generic/verifier/verifier_utils.py

```diff
@@ -0,0 +1,171 @@
+from app.facades.trust_registry import (
+    Actor,
+    actor_by_did,
+    get_trust_registry,
+)
+from app.generic.verifier.facades.acapy_verifier import Verifier
+from app.error.cloud_api_error import CloudApiException
+from app.facades.acapy_wallet import assert_public_did
+
+logger = logging.getLogger(__name__)
+
+
+# VERIFIER
+async def check_tr_for_verifier(
+    aries_controller: AcaPyClient,
+    proof_request,
+) -> Optional[bool]:
+    """Check transaction requirements against trust registry for verifier"""
+    # 1. Check agent has public did
+    # CASE: Agent has public DID
+    try:
+        pub_did = await assert_public_did(aries_controller=aries_controller)
+    except Exception:
+        # CASE: Agent has NO public DID
+        ## check via connection -> invitation key
+        connection_record = await get_connection_record(
+            aries_controller=aries_controller,
+            connection_id=proof_request.connection_id,
+        )
+        # get invitation key
+        invitation_key = connection_record.invitation_key
+        pub_did = f"did:sov:{invitation_key}"
+    # Try get actor from TR
+    actor = await get_actor(did=pub_did)
+    # 2. Check actor has role verifier, raise exception otherwise
+    if not is_verifier(actor=actor):
+        raise CloudApiException(
+            f"{actor['name']} is not a valid verifier in the trust registry."
+        )
+    else:
+        return True
+
+
+# PROVER
+async def check_tr_for_prover(
+    aries_controller: AcaPyClient, prover: Verifier, proof_request
+) -> Optional[bool]:
+    """Check transaction requirements against trust registry for prover"""
+    # get connection record
+    proof_id = proof_request.proof_id
+    connection_id = await get_connection_from_proof(
+        aries_controller=aries_controller, proof_id=proof_id, prover=prover
+    )
+    connection_record = await get_connection_record(
+        aries_controller=aries_controller,
+        connection_id=connection_id,
+    )
+    # TODO: (In other PR) handle case where no conneciton id exists
+    # instead of simply rejecting the request
+    if not connection_record.connection_id:
+        raise CloudApiException(f"Cannot proceed. No connection ID", 404)
+
+    # Case 1: connection NOT made with publid DID
+    if connection_record.their_public_did and connection_record.their_public_did != "":
+        pub_did = f"did:sov:{connection_record.their_public_did}"
+    # Case 2: connection made with public DID
+    else:
+        invitation_key = connection_record.invitation_key
+        pub_did = ed25519_verkey_to_did_key(key=invitation_key)
+
+    # Try get actor from TR
+    actor = await get_actor(did=pub_did)
+    # 2. Check actor has role verifier
+    if not is_verifier(actor=actor):
+        raise CloudApiException("Could not verify prover against trust registry", 401)
+
+    # Get schema ids
+    schema_ids = await get_schema_ids(aries_controller=aries_controller, prover=prover)
```

Comment:
This checks if the credentials the user has, have schema ids that  are valid in the trust request. However this doesn't actually test anything for the proof request.

The user could have a credential in its wallet with a different schema id. The thing is the verifier should only be allowed to ask for a request for one of the schemas registered in the trust registry. This can be quite hard as proof request can be defined quite dynamically.  I think we're left with two options:

1. require all restrictions on requested_attributes and request_predicates to require the schema_id to be present
2. fetch all credentials that could fulfil the request (instead of fetching _all_ credentials )and filter out credentials that don't have a correct schema id OR throw an error if any credential is in the list with a schema id not in the trust registry.

I think 1. is easiest.

Reply:
alright will change that

Reply:
actually upon second thought. maybe I'm still having this somewhat wrong in my mental model of this, but what I've done is get the schemas for the request and then check whether they are all in the trust registry (the lines of code below)

```python
if await is_valid_schemas(schema_ids=schema_ids):
```

Not sure I'm missing sth or potentially making a mistake in my thinking? Thought that should cover your remarks?!